paddle/fluid/framework/ir/cost_model.cc

```diff
@@ -0,0 +1,250 @@
+      VLOG(4) << (*time_events)[i][j].name() << " "
+              << (*time_events)[i][j].attr() << std::endl;
+      VLOG(4) << "This: " << &(*time_events)[i][j]
+              << ", Parent: " << (*time_events)[i][j].parent() << std::endl;
+      if ((*time_events)[i][j].role() == platform::EventRole::kInnerOp) {
+        VLOG(4) << "role kInnerOp" << std::endl;
+      } else if ((*time_events)[i][j].role() ==
+                 platform::EventRole::kUniqueOp) {
+        VLOG(4) << "role kUniqueOp" << std::endl;
+      } else if ((*time_events)[i][j].role() ==
+                 platform::EventRole::kOrdinary) {
+        VLOG(4) << "role kOrdinary" << std::endl;
+      } else if ((*time_events)[i][j].role() == platform::EventRole::kSpecial) {
+        VLOG(4) << "role kSpecial" << std::endl;
+      }
+
+      if ((*time_events)[i][j].type() == platform::EventType::kPopRange) {
+        VLOG(4) << "type kPopRange" << std::endl;
+      } else if ((*time_events)[i][j].type() ==
+                 platform::EventType::kPushRange) {
+        VLOG(4) << "type kPushRange" << std::endl;
+      } else if ((*time_events)[i][j].type() == platform::EventType::kMark) {
+        VLOG(4) << "type kMark" << std::endl;
+      }
+      VLOG(4) << std::endl;
+    }
+  }
+
+  for (size_t i = 0; i < mem_events->size(); ++i) {
+    for (size_t j = 0; j < (*mem_events)[i].size(); ++j) {
+      VLOG(4) << "Print mem event (" << i << ", " << j << ")" << std::endl;
+      VLOG(4) << (*mem_events)[i][j].annotation() << std::endl;
+    }
+  }
+}
+
+std::string ToLowerCopy(const std::string& in) {
+  std::string out(in);
+  std::transform(out.begin(), out.end(), out.begin(),
+                 [](unsigned char c) { return std::tolower(c); });
+  return out;
+}
+
+CostData CostModel::ProfileMeasure(
+    const ProgramDesc& main_program, const ProgramDesc& startup_program,
+    const std::string& device,
+    const std::vector<std::string>& fetch_cost_list) const {
+  // Currently fetch_cost_list is useless
+  // TODO(zhhsplendid): support different fetch data
+
+  platform::ProfilerState profiler_state;
+  platform::Place place;
+
+  std::string device_lower_case = ToLowerCopy(device);
+  if (device_lower_case == "cpu") {
+    profiler_state = platform::ProfilerState::kCPU;
+    place = platform::CPUPlace();
+  } else if (device_lower_case == "gpu") {
+    profiler_state = platform::ProfilerState::kAll;
+    place = platform::CUDAPlace();
+  } else {
+    PADDLE_THROW(platform::errors::Unimplemented(
+        "Not support %s in CostModel now", device));
+  }
+
+  Executor executor(place);
+  Scope scope;
+  executor.Run(startup_program, &scope, /*block_id = */ 0);
+
+  // TODO(zhhsplendid): handle the case that Profiler is already enabled
+  SetTracerOption(platform::TracerOption::kAllOpDetail);
+  EnableProfiler(profiler_state);
+  executor.Run(main_program, &scope, /*block_id = */ 0);
+
+  std::vector<std::vector<Event>>* time_events =
```

Comment:
I think the best idea is using `unique_ptr` or `shared_ptr` rather than raw pointer.

Reply:
Fine to me, change done.